In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
data =pd.read_csv(r"D:\DATA SCIENCE\TASKS\student-data.csv")
data.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  internet romantic  famrel  freetime  goout Dalc Walc health absences passed  
0       no       no       4         3      4    1    1      3        6     no  
1      yes       no       5         3      3    1    1      3        4     no  
2      yes       no       4         3      2    2    3      3       10    yes  
3      yes      yes       3         2      2    1    1      5        2    yes  
4       no       no       4         3      2    1    2      5        4    yes  

[5 rows x 31 columns]

In [3]:
data.shape

(395, 31)

In [4]:
data.describe()

age        Medu        Fedu  traveltime   studytime    failures  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean    16.696203    2.749367    2.521519    1.448101    2.035443    0.334177   
std      1.276043    1.094735    1.088201    0.697505    0.839240    0.743651   
min     15.000000    0.000000    0.000000    1.000000    1.000000    0.000000   
25%     16.000000    2.000000    2.000000    1.000000    1.000000    0.000000   
50%     17.000000    3.000000    2.000000    1.000000    2.000000    0.000000   
75%     18.000000    4.000000    3.000000    2.000000    2.000000    0.000000   
max     22.000000    4.000000    4.000000    4.000000    4.000000    3.000000   

           famrel    freetime       goout        Dalc        Walc      health  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean     3.944304    3.235443    3.108861    1.481013    2.291139    3.554430   
std      0.896659    0.998862    1.113278    0.890741    1.287897    1.390303   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      4.000000    3.000000    2.000000    1.000000    1.000000    3.000000   
50%      4.000000    3.000000    3.000000    1.000000    2.000000    4.000000   
75%      5.000000    4.000000    4.000000    2.000000    3.000000    5.000000   
max      5.000000    5.000000    5.000000    5.000000    5.000000    5.000000   

         absences  
count  395.000000  
mean     5.708861  
std      8.003096  
min      0.000000  
25%      0.000000  
50%      4.000000  
75%      8.000000  
max     75.000000

In [5]:
data.isna().sum()

school        0
sex           0
age           0
address       0
famsize       0
Pstatus       0
Medu          0
Fedu          0
Mjob          0
Fjob          0
reason        0
guardian      0
traveltime    0
studytime     0
failures      0
schoolsup     0
famsup        0
paid          0
activities    0
nursery       0
higher        0
internet      0
romantic      0
famrel        0
freetime      0
goout         0
Dalc          0
Walc          0
health        0
absences      0
passed        0
dtype: int64

In [6]:
data.dtypes

school        object
sex           object
age            int64
address       object
famsize       object
Pstatus       object
Medu           int64
Fedu           int64
Mjob          object
Fjob          object
reason        object
guardian      object
traveltime     int64
studytime      int64
failures       int64
schoolsup     object
famsup        object
paid          object
activities    object
nursery       object
higher        object
internet      object
romantic      object
famrel         int64
freetime       int64
goout          int64
Dalc           int64
Walc           int64
health         int64
absences       int64
passed        object
dtype: object

In [7]:
#checking the unique values
data_obj=[f for f in data.columns if data[f].dtypes=='O']
for d in data_obj:
    print(d,data[d].unique())

school ['GP' 'MS']
sex ['F' 'M']
address ['U' 'R']
famsize ['GT3' 'LE3']
Pstatus ['A' 'T']
Mjob ['at_home' 'health' 'other' 'services' 'teacher']
Fjob ['teacher' 'other' 'services' 'health' 'at_home']
reason ['course' 'other' 'home' 'reputation']
guardian ['mother' 'father' 'other']
schoolsup ['yes' 'no']
famsup ['no' 'yes']
paid ['no' 'yes']
activities ['no' 'yes']
nursery ['yes' 'no']
higher ['yes' 'no']
internet ['no' 'yes']
romantic ['no' 'yes']
passed ['no' 'yes']


In [8]:
data_obj.pop()

'passed'

In [9]:
data['passed']=data['passed'].map({'yes':1,'no':0})

In [10]:
X=data.drop(['passed'],axis=1)
y=data['passed']

In [12]:
from sklearn.preprocessing import LabelEncoder

In [16]:
lab_en = LabelEncoder()
l1 = ['school', 'sex', 'address', 'famsize', 'Pstatus', 'schoolsup', 'famsup',
       'paid', 'activities', 'Mjob','Fjob', 'nursery', 'higher', 'internet', 'romantic']
for x in l1:
    data[x] = lab_en.fit_transform(data[x])

In [17]:
data.head()

school  sex  age  address  famsize  Pstatus  Medu  Fedu  Mjob  Fjob  ...  \
0       0    0   18        1        0        0     4     4     0     4  ...   
1       0    0   17        1        0        1     1     1     0     2  ...   
2       0    0   15        1        1        1     1     1     0     2  ...   
3       0    0   15        1        0        1     4     2     1     3  ...   
4       0    0   16        1        0        1     3     3     2     2  ...   

  internet romantic  famrel  freetime  goout  Dalc  Walc  health  absences  \
0        0        0       4         3      4     1     1       3         6   
1        1        0       5         3      3     1     1       3         4   
2        1        0       4         3      2     2     3       3        10   
3        1        1       3         2      2     1     1       5         2   
4        0        0       4         3      2     1     2       5         4   

   passed  
0       0  
1       0  
2       1  
3       1  
4       1  

[5 rows x 31 columns]

In [18]:
data_obj=[f for f in X.columns if X[f].dtypes=='O']
X=pd.get_dummies(X,columns=data_obj,drop_first=True)
X.head()

age  Medu  Fedu  traveltime  studytime  failures  famrel  freetime  goout  \
0   18     4     4           2          2         0       4         3      4   
1   17     1     1           1          2         0       5         3      3   
2   15     1     1           1          2         3       4         3      2   
3   15     4     2           1          3         0       3         2      2   
4   16     3     3           1          2         0       4         3      2   

   Dalc  ...  guardian_mother  guardian_other  schoolsup_yes  famsup_yes  \
0     1  ...                1               0              1           0   
1     1  ...                0               0              0           1   
2     2  ...                1               0              1           0   
3     1  ...                1               0              0           1   
4     1  ...                0               0              0           1   

   paid_yes  activities_yes  nursery_yes  higher_yes  internet_yes  \
0         0               0            1           1             0   
1         0               0            0           1             1   
2         1               0            1           1             1   
3         1               1            1           1             1   
4         1               0            1           1             0   

   romantic_yes  
0             0  
1             0  
2             0  
3             1  
4             0  

[5 rows x 39 columns]

#### SPLITTING TO TRAIN AND TEST

In [19]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,random_state=40,test_size=.35) #35% data is for testing

### LOGISTIC REGERESSION MODEL

In [20]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score
log_model=LogisticRegression(solver='newton-cg')
log_model.fit(X_train,y_train)
prediction=log_model.predict(X_test)
print("F1_score :",f1_score(y_test,prediction))
print("Accuracy score :",accuracy_score(y_test,prediction))
print("Confusion_matrix :",confusion_matrix(y_test,prediction))

F1_score : 0.8078817733990148
Accuracy score : 0.7194244604316546
Confusion_matrix : [[18 22]
 [17 82]]


### RANDOM FOREST CLASSIFIER

In [21]:
from sklearn.ensemble import RandomForestClassifier
rf_mod = RandomForestClassifier(random_state=40)
rf_mod.fit(X_train,y_train)
y_pred=rf_mod.predict(X_test)
print('Accuracy Score:',accuracy_score(y_test,y_pred))
print('Conclusion matrix\n',confusion_matrix(y_test,y_pred))

Accuracy Score: 0.7482014388489209
Conclusion matrix
 [[14 26]
 [ 9 90]]


### SVM MODEL

In [22]:
from sklearn.svm import SVC
svm_gauss=SVC(kernel='rbf')
svm_gauss.fit(X_train,y_train)
y_pred=svm_gauss.predict(X_test)

In [23]:
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report
print('Accuracy Score:',accuracy_score(y_test,y_pred))
print('Conclusion matrix\n',confusion_matrix(y_test,y_pred))

Accuracy Score: 0.7050359712230215
Conclusion matrix
 [[ 1 39]
 [ 2 97]]


# RANDOM FOREST CLASSIFIER MODEL GIVES BETTER ACCURACY AT 74.8%